text/0010-design-libraries.md

```diff
@@ -0,0 +1,47 @@
+# The Styleguide and Design Libraries
+- Start Date: 2019/06/08
+- RFC PR: https://github.com/openmrs/openmrs-rfc-frontend/pull/10
+
+## Decision, including impact on distributions
+The [OpenMRS styleguide](/text/0008-styleguide.md) does not use a third party design library's code (such as Twitter Bootstrap,
+Material, Semantic UI). Instead, the styleguide is built from scratch in collaboration with UX designers in the OpenMRS community.
+UX principles and design elements are liberally borrowed from other design systems and libraries, but the code for the
+OpenMRS styleguide is written by the OpenMRS community.
+
+Distributions should use the OpenMRS styleguide as the primary source of making a consistent user experience, look and feel,
+layout, and base-level components. In addition to the OpenMRS styleguide, distributions may use design libraries if they prefer to.
+However, it is not necessary (nor encouraged) to do so.
+
+## Definition
+A design library is a third-party javascript library that provides pre-written css and javascript components. Design libraries
+are installed into code and are often used by the frontend community for base-level components (such as buttons, form elements,
+and tooltips).
+
+At the time of this writing, [Material UI](https://material-ui.com/) (from Google), [Twitter Bootstrap](https://getbootstrap.com/),
+and [Semantic UI](https://semantic-ui.com/) are a few of the popular design libraries.
+
+A custom build of a design library is a customized version of such a library. Customizations often include things like custom
+css class naming, additional css class definitions, and custom branding variables. Design libraries often provide a
+well-supported way to create a custom build of their library. Organizations can use custom builds instead of the generic version
+of the library to make the design library their own.
+
+## Reason for decision
+- Upgrading design libraries is often very difficult because breaking changes span across an organization's entire
```

Comment:
Since microfrontends can manage their styling independently, doesn't that enable us to upgrade versions of a library piecemeal across the different components? Maybe some microfrontends will get left behind, but we've explicitly chosen an architecture in which big rewrites are unlikely to be necessary.

Reply:
Yes, exactly. The styleguide, though, isn't independent from the other microfrontends, which means that breaking changes to the styleguide impact all of the microfrontends at once.

Here's an example situation that could happen if the styleguide is material-ui:
1) Styleguide uses material-ui@4.1.0, since that's the current latest version.
2) All microfrontends use the css classes exposed by material-ui
3) In 1-2 years, material-ui@5 comes out. It has some breaking changes to the exposed css classes that impact some, but not all, microfrontends.
4) What do we do with the styleguide?
a) We decide not to upgrade styleguide to material-ui@5 until all the important microfrontends are forwards-compatible with material-ui@5. It is logged as technical debt.
b) OR: we decide to upgrade styleguide to material-ui@5 immediately, encouraging all microfrontends to upgrade as soon as possible.

I don't like either of the outcomes very much. 4a often results in the styleguide remaining on material-ui@4 for a very long time. Compare this to OpenMRS recently getting to Java 8 after what I understand to have been a significant effort, even though Java 12 is the current version. Or another one is how much of OpenMRS frontend uses jquery@1 (at least the few files i've looked at), even though the latest version is jquery 3.

For 4b, I've seen immediate upgrades to a core library like that result in people staying on older versions of the core library indefinitely because they never get around to material-ui forwards compatibility. For example, my understanding is that Kenya EMR is still on OpenMRS@1 because upgrading would be hard.

^^ All of this just goes to show the importance of backwards compatibility with shared, core code. And I believe backwards compatibility to be far easier to accomplish if you have 1) a minimal API surface (in this case, less css classes and components) and 2) control over the code that needs to be backwards compatible.

Reply:
Oh, wow, I understand now. Yep.

Reply:
But wait, wouldn't our library be subject to the same constraints? Is there a big difference between not being allowed to upgrade from Material UI 4 to Material UI 5, and not being allowed to make breaking changes in our own library?

Reply:
> Is there a big difference between not being allowed to upgrade from Material UI 4 to Material UI 5, and not being allowed to make breaking changes in our own library?

Nope. No difference at all. A breaking change to a homegrown styleguide is just as bad as a breaking change to a third party styleguide. Which is why I think we should keep core code as backwards compatible as possible. And that writing and maintaining that code ourselves is the best way to achieve that.

From my earlier comment:
> I believe backwards compatibility to be far easier to accomplish if you have 1) a minimal API surface (in this case, less css classes and components) and 2) control over the code that needs to be backwards compatible.

Reply:
I'm going to update the PR to add a bullet point about this ^^ --  maintaining backwards compatibility wherever possible in the styleguide is important, and writing our own code facilitates that imo.

Reply:
Nice. Ok, that's quite compelling.

Reply:
In the OpenMRS community, there has been a general drive to run away from our custom things that we have to maintain, force new comers to learn, to get towards such things that new developers are already familiar with from other projects and which we do not have to maintain ourselves. Should we bring this to a wider audience on talk to get what others think? https://wiki.openmrs.org/display/RES/2017-01-23+Design+Forum

Reply:
Could you explain your reservations more, @dkayiwa? Do you prefer using a customized build of Twitter Bootstrap / similar? Or something else entirely?

Reply:
This was not me. Was mainly re-echoing what i have seen discussed.
cc @ssmusoke @bmamlin